lib/grp.gi

```diff
@@ -1013,6 +1013,32 @@ RedispatchOnCondition( FittingSubgroup, true, [IsGroup], [IsFinite], 0);
 ##
 #M  FrattiniSubgroup( <G> ) . . . . . . . . . .  Frattini subgroup of a group
 ##
+InstallMethod( FrattiniSubgroup, "generic method for nilpotent groups",
+            [ IsGroup ],
+            RankFilter( IsGroup and IsNilpotentGroup )
+            - RankFilter( IsGroup ),
+function(G)
+local i, p, q, gen, Gf;
+    if IsTrivial(G) then
+      return G;
+    elif IsAbelian(G) then
+        gen := [ ];
+        for i in [1..Length(AbelianInvariants(G))] do
+            q := AbelianInvariants(G)[i];
+            if q<>0 and not IsPrime(q) then
+                p := SmallestRootInt(q);
+                Add(gen, IndependentGeneratorsOfAbelianGroup(G)[i]^p);
+            fi;
```

Comment:
Calling `AbelianInvariants` and `IndependentGeneratorsOfAbelianGroup` again and again like that makes me cringe. While we cache these values, each of these calls incurs an overhead, as the system has to perform a lookup in a component object. So please store those into temporary variables.